numpy/core/tests/test_indexing.py

```diff
@@ -511,6 +511,98 @@ def test_indexing_array_negative_strides(self):
+        assert_equal(as_index_tuple(ell),  (ell,))
+        assert_equal(as_index_tuple(None), (None,))
+        assert_equal(as_index_tuple(sl),   (sl,))
+        assert_equal(as_index_tuple(arr),  (arr,))
+        assert_equal(as_index_tuple(obj),  (obj,))
+
+        # tuples are untouched
+        assert_equal(as_index_tuple((1, 2, 3)),    (1, 2, 3))
+        assert_equal(as_index_tuple((1, 2, ell)),  (1, 2, ell))
+        assert_equal(as_index_tuple((1, 2, None)), (1, 2, None))
+        assert_equal(as_index_tuple((1, 2, sl)),   (1, 2, sl))
+        assert_equal(as_index_tuple((1, 2, arr)),  (1, 2, arr))
+
+        # sequences of scalars are wrapped
+        assert_equal(as_index_tuple([1, 2, 3]), ([1, 2, 3],))
+
+        # sequences containing slice objects or ellipses are tuple-ified
+        assert_equal(as_index_tuple([1, 2, ell]),  (1, 2, ell))
+        assert_equal(as_index_tuple([1, 2, None]), (1, 2, None))
+        assert_equal(as_index_tuple([1, 2, sl]),   (1, 2, sl))
+        assert_equal(as_index_tuple([1, 2, arr]),  (1, 2, arr))
+
+        # unless they are >= np.MAXDIMS, in which case they are always wrapped
+        nd = np.MAXDIMS
+        assert_equal(as_index_tuple(nd * [1]),    (nd * [1],))
+        assert_equal(as_index_tuple(nd * [ell]),  (nd * [ell],))
+        assert_equal(as_index_tuple(nd * [None]), (nd * [None],))
+        assert_equal(as_index_tuple(nd * [sl]),   (nd * [sl],))
+        assert_equal(as_index_tuple(nd * [arr]),  (nd * [arr],))
+
+    def test_as_index_tuple_broken_getitem(self):
+        from numpy.core.multiarray import as_index_tuple
+
+        # test sequences with a broken __getitem__
+        def make_broken_sequence(base, items):
+            class Broken(base):
+                def __len__(self):
+                    return len(items)
+                def __getitem__(self, i):
+                    val = items[i]
+                    if isinstance(val, Exception):
+                        raise val
+                    return val
+            return Broken()
+
+        # error comes first, so just treat as a scalar
+        idx = make_broken_sequence(object, [1, ValueError(), None])
+        assert_raises(ValueError, operator.getitem, idx, 1)
+        assert_equal(as_index_tuple(idx),  (idx,))
+
+        # none comes first, so commit to making the tuple
+        idx = make_broken_sequence(object, [1, None, ValueError()])
+        assert_raises(ValueError, operator.getitem, idx, 2)
+        assert_raises(ValueError, as_index_tuple, idx)
+
+        # tuples subclasses error in both cases
+        idx = make_broken_sequence(tuple, [1, ValueError(), None])
+        assert_raises(ValueError, operator.getitem, idx, 1)
+        assert_raises(ValueError, as_index_tuple, idx)
+
+        idx = make_broken_sequence(tuple, [1, None, ValueError()])
+        assert_raises(ValueError, operator.getitem, idx, 2)
+        assert_raises(ValueError, as_index_tuple, idx)
```

Comment:
This and the test above it is a behaviour change - previously, `idx.__getitem__` would never be called at all, and `idx` would be treated as if it were `()`.